Add full support for Android Emulator 2.0

redeamer opened the pull request:
Currently the plugin does not support the 'new' Android Emulator 2.0, as (introduced in SDK Tool v25) the first versions did not support all needed flags and parameters needed by the plugin. As result, the classic/legacy engine was used for v25 and above.
Since SDK Tools v26 (I think even since 25.0.3) the new emulator engine supports all needed flags and parameters.
I added support for the 'new' emulator engine by respecting the new location and usage of tools (emulator, new avdmanager) fully introduced in v25.3.0 and remove the classic flag on v26 and above. Fixing the following issues: [JENKINS-28941](https://issues.jenkins-ci.org/browse/JENKINS-28941), [JENKINS-39097](https://issues.jenkins-ci.org/browse/JENKINS-39097), [JENKINS-40178](https://issues.jenkins-ci.org/browse/JENKINS-40178), [JENKINS-43215](https://issues.jenkins-ci.org/browse/JENKINS-43215), [JENKINS-43557](https://issues.jenkins-ci.org/browse/JENKINS-43557), [JENKINS-44490](https://issues.jenkins-ci.org/browse/JENKINS-44490)

The plugin creates AVDs and starts the emulator in legacy mode for SDK Tools v25.2.0 and below, for v25.3.0 and up the new tools are used, but the classic option is used (as before) and starting with v26 emulator is used.

The following changes where needed/made:
- To respect the new avdmanager a new configuration option was added ('Device definition') which is an optional parameter. It is used to define a specific device ('avdmanager list device'). The value is only respected if SDK Tools >= 25.3.0 are used.
- A new ToolLocator (EmulatorToolLocator) were added to respect the moved emulator binaries ('tools'-dir vs 'emulator'-dir)
- Added a helper class AndroidPlatformVersions which holds the current data of all known Android-Platforms. The static information of AndroidPlatform were moved here. AndroidPlatform itself got refactored to handle and retrieve more information needed for the emulator (package path) and
for more convenient access to vendor and api information of the used platform (platform parsing, eg: "Google Inc.:Google APIs:24").
- SdkInstaller: using AndroidPlatform-Helper instead of parsing platform-String
- Added tests for new functionality (not true for SdkInstaller, had no tests before)
- Fixed Android Emulator 2.0 Bug ([Android Issue 64356053](https://issuetracker.google.com/issues/64356053)) where '-noaudio', '-no-audio' and '-audio none' does not work. This needs to be set via the Environment variables. So if one of those parameters is detected to start the emulator, the Env-Var 'QEMU_AUDIO_DRV=none' is set.
- While fixing the audio issue above, all env-variable-names were moved to Constants.java

Known Issues:
- `-engine classic` removed for SDK Tools v25 on request of maintainer, will be added to the Known issues section/changelog, that users experience problems with that version should set the cli-args manually
- Snapshots ('-a') are not supported with the qemu2 engine, so the setting is completely ignored vor SDK Tools v25 and up

redeamer commented:
Updated description a little bit and added the 'changes'-list (which was not complete before) and added the 'known issues/possible improvements' section.
I would really love to get input or (in an ideal scenario) even a code review for this change.

Sergio-Mira commented:
Looks good to me!

redeamer commented:
Fixed another issue [(Android Issue 64356053)](https://issuetracker.google.com/issues/64356053) with the Qemu2 emulator: The CLI-options to disable audio '-noaudio', '-no-audio' and '-audio none' does not work. This needs to be set via the Environment variables. So if one of those parameters is detected to start the emulator, the Env-Var 'QEMU_AUDIO_DRV=none' is set.
While implementing that change, I moved all used environment variable names to the Constants.java-File.

cl1ent commented:
hey @redeamer,

I've compiled the plugin with your commits but for some reason it still runs some legacy commands. is this intended? an example is:

> 11:00:08 $ /usr/local/android-sdk/tools/android list target
11:00:09 [android] The configured Android platform needs to be installed: android-24
11:00:09 $ /usr/local/android-sdk/tools/android list target
11:00:10 $ /usr/local/android-sdk/tools/android list target
11:00:11 [android] Installing the 'sys-img-x86-android-24' SDK component(s)...
11:00:11 $ /usr/local/android-sdk/tools/android update sdk -u -a -t sys-img-x86-android-24
11:00:11 *************************************************************************
11:00:11 The "android" command is deprecated.
11:00:11 For manual SDK, AVD, and project management, please use Android Studio.
11:00:11 For command-line tools, use tools/bin/sdkmanager and tools/bin/avdmanager
11:00:11 *************************************************************************
11:00:31 Timed out waiting for input.
11:00:31 To suppress this prompt, run with --use-sdk-wrapper or set USE_SDK_WRAPPER.
11:00:31 [android] Using Android SDK: /usr/local/android-sdk

orrc commented:
I noticed this as well, and creating an emulator failed for me.  Wasn't sure whether it was intentional.

Another quick note from some brief testing last week: I think if no target ABI is entered, a `NullPointerException` occurs.

redeamer commented:
Hi,
thanks for the input, I'll check and fix that 'target-ABI NullPointer and I noticed the creating an emulator failed issue already. As the sdk-update parts worked at first glance on my machine, by updating from an older version, I left them untouched. But it reproducible fails with a empty sdk-root where the currently defined SDK gets download. I'm not sure if the 'upgrade' from different older sdk's let always to the same resulting updated SDK or, more likely my initial used SDK dir was somehow broken.
I'm currently working on splitting the SDK install and upgrade commands to work with differnent SDK versions and the up-to-date commands. It works already on my local machine but the code needs cleanup and the unit-tests are not yet all passed.
I think I'm able to commit the change by the evening or tomorrow morning (CEST) with the following improvements:
- Download the latest SDK if non is exisiting and Plugin shall handle the sdk-root
- Use sdkmanager everywhere with new SDK
- Having unit-tests for the commands which gets executed (sdkmanager, avdmanager, adb,...)

orrc commented:
FYI: I've added a `Jenkinsfile` to this repo, so anybody who wants to test a build can download the latest from here by choosing the most recent run, then going to the Artifacts tab:
https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fandroid-emulator-plugin/activity?branch=PR-71

durgasunil86 commented:
Hi @redeamer , 
I tried compiling the plugin with your commits but am getting a null pointer exception while launching the simulator. I am trying to launch an existing emulator by passing in the AVD name in the configuration. 

Console Log : 

Checking out Revision d0aa6d7c4ecca377b5016c7abd99f25c04d2dd6c (refs/remotes/origin/master)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f d0aa6d7c4ecca377b5016c7abd99f25c04d2dd6c
 > git rev-list d0aa6d7c4ecca377b5016c7abd99f25c04d2dd6c # timeout=10
Allocating TCP port 5790
  -> Assigned 5790
TCP port allocation complete
[android] Using Android SDK: /Users/jenkins/slave/tools/android-sdk
[android] Waiting 10 seconds before starting emulator...
$ /Users/jenkins/slave/tools/android-sdk/platform-tools/adb start-server
 daemon not running. starting it now at tcp:5863 
 daemon started successfully 
$ /Users/jenkins/slave/tools/android-sdk/platform-tools/adb start-server
FATAL: null
java.lang.NullPointerException
	at hudson.plugins.android_emulator.EmulatorConfig.getCommandArguments(EmulatorConfig.java:362)
	at hudson.plugins.android_emulator.AndroidEmulator.doSetUp(AndroidEmulator.java:344)
	at hudson.plugins.android_emulator.AndroidEmulator.setUp(AndroidEmulator.java:275)
	at hudson.model.Build$BuildExecution.doRun(Build.java:157)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:534)
	at hudson.model.Run.execute(Run.java:1728)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:405)
Archiving artifacts
[htmlpublisher] Archiving HTML reports...
[htmlpublisher] Archiving at BUILD level /Users/jenkins/slave/workspace/android_slideshow_UITests/slideshow/app/build/reports/androidTests/connected to /var/lib/jenkins/jobs/android_slideshow_UITests/builds/1033/htmlreports/HTML_Report
Recording test results
ERROR: Step ‘Publish JUnit test result report’ failed: Test reports were found but none of them are new. Did tests run?

redeamer commented:
- I have moved the sdk tool commands and the adb shell commands to seperate classes (android_emulator.sdk.cli package) and having a factory to retrieve the correct implementation for the installed sdk tools version (sdk tools commands) or target-api (adb shell commands).
I tested the calls with and old sdk installation (r24), a current manual installed SDK and with an empty dir (current auto install). All scenarios worked, by I'm not sure, if there are other setups where the sdk handling will not work. **So I would advice to backup the current 'android-sdk'-directory.** 
- The NPE expection @durgasunil86 mentioned should be fixed (-skin option is not used, if no valid screenresolution is given)
- The NPE if no abi is given is not yet resolved: as i got it, it is not possible to start the emulator without specifying the 'Package path' (-k), which is in form of '`system-images;<target>;<tag>;<abi/arch>`'. I will fix the NPE if no abi is specified and add a warning to the config page that it needs to be given using the 'new' emulator.

redeamer commented:
NullPointerException is fixed if no ABI is given and avdmanager is used to create the AVD.
An error is printed and an EmulatorCreationException is thrown if no ABI is given in conjunction with the new tools.
Additionally I added a warning to the config page and updated the ABI description.

durgasunil86 commented:
Hi @redeamer , I tried the latest commits and have couple issues to report. 

1.  Now getting the below error , when i disable the option to _'Automatically install Android components when required'_ at the global settings. I did set the environment variables for ANDROID_SDK_HOME, and ANDROID_HOME. Not sure why the plugin is not able to find the path for tools.
           [android] Required Android tools not found in PATH; cannot continue

I have set the below two variables and i believe this should be enough, from what i glanced from the plugin code.
export ANDROID_HOME=/Users/jenkins/slave/tools/android-sdk
export ANDROID_SDK_HOME=/Users/jenkins/slave/tools/android-sdk

2. However if i enable the '_Automatically install Android components when required_' option then there is a null reference exception.

android] No Android SDK found; let's install it automatically...
FATAL: null
java.lang.NullPointerException
	at hudson.plugins.android_emulator.util.Utils.getToolCommand(Utils.java:430)
	at hudson.plugins.android_emulator.util.Utils.runAndroidTool(Utils.java:476)
	at hudson.plugins.android_emulator.util.Utils.runAndroidTool(Utils.java:469)
	at hudson.plugins.android_emulator.util.Utils.runAndroidTool(Utils.java:463)
	at hudson.plugins.android_emulator.SdkInstaller.isPlatformInstalled(SdkInstaller.java:323)
	at hudson.plugins.android_emulator.SdkInstaller.installPlatform(SdkInstaller.java:269)
	at hudson.plugins.android_emulator.SdkInstaller.installDependencies(SdkInstaller.java:247)
	at hudson.plugins.android_emulator.AndroidEmulator.setUp(AndroidEmulator.java:271)
	at hudson.model.Build$BuildExecution.doRun(Build.java:157)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:534)
	at hudson.model.Run.execute(Run.java:1728)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:405)
Archiving artifacts
[htmlpublisher] Archiving HTML reports...
[htmlpublisher] Archiving at BUILD level /Users/jenkins/slave/workspace/android_slideshow_unitTests/slideshow/app/build/reports/androidTests/connected to /var/lib/jenkins/jobs/android_slideshow_unitTests/builds/1362/htmlreports/HTML_Report
Recording test results
ERROR: Step ‘Publish JUnit test result report’ failed: Test reports were found but none of them are new. Did tests run?

redeamer commented:
Hi @durgasunil86,
there seems to be a problem with the SDK root detection/validation.
Would it be possible to send me the contents of the following directories:
`ls -l /Users/jenkins/slave/tools/android-sdk`
`ls -l /Users/jenkins/slave/tools/android-sdk/tools`
`ls -l /Users/jenkins/slave/tools/android-sdk/platform-tools`

If you do not want to post (even only a standard android SDK) local directory contents, you can send the output privately via my gmail<single_dot>com address starting with michael.musenbrock@g... (I hope this is understandable and still confusing for bots ;) )

redeamer commented:
Hi @durgasunil86,
thanks for the directory-content, they looked as expected.
The error should now be fixed. The issue was due to running on a slave machine, as I added a log-statement in a MasterToSlaveCallable, where the used PrintStream is not serializable and the exception was suppressed in the original code.
I removed the newly added Log-statement but added logs for the Exceptions which get catched in getAndroidSdk-method. No I can run my jobs without problems on a slave machine.

durgasunil86 commented:
Hi @redeamer  ,  
I was able to build the latest commit and successfully tested the plugin. It works great without any issues !!  Thanks a lot for taking up this and fixing all the issues. The branch is good to merge from my side.

redeamer commented:
Hi @orrc,
there are still two questions remaining:
- Should I add a config option to use the -classic switch, and defaulting to 'true', to not automatically changing the behavior on plugin-upgrade. But the more i think of, the less ideal is it: new installs would still use classic so we should just point it out in the changelog.
- The 'sdkmanager' offers a dedicated '--update' switch. Is this wanted to be executed if we handle the SDK installation by the plugin? Even if that means, a future upgrade may break the compatibility?

br michael

orrc commented:
Hey, greetings from Jenkins World! 😄 

> Should I add a config option to use the `-classic` switch

I think fewer config options are better. Probably the plugin should stop specifying the `-engine` flag altogether, allowing it to default to `auto`. If someone really wants to override it, they can add `-engine classic` to the "additional arguments" field.

The 2.0 engine should work well enough now that even for upgrading users, we can let the automatic behaviour of the emulator be used.  So yeah, we can just mention this in the release notes, and that people can override with `-engine classic` if they really need.

> The 'sdkmanager' offers a dedicated '--update' switch. Is this wanted to be executed if we handle the SDK installation by the plugin?

For now, it's probably preferable not to call `sdkmanager --update`, since things can break between SDK Tools versions.

redeamer commented:
Hey,

as the plugin won't start the emulator 2.0 in v25 without the switch, I think it's okay to keep the flag there for this particular version. In v26 the '-engine' switch is not used anyways by the plugin. And the update will break the emulator for existing v25 users. But my opinion is not strong on this one ;) So If I don't hear any concerns or opinions in the next days, I will simply remove the option.

slightly jealous greetings from rainy Central Europe to SF ;)

cl1ent commented:
+1 merge

orrc commented:
Greetings from rainy and jetlagged Central Europe! 👋 

> as the plugin won't start the emulator 2.0 in v25 without the switch, I think it's okay to keep the flag there for this particular version. In v26 the '-engine' switch is not used anyways by the plugin.

We should be able to remove the use of `-engine` entirely from the code, as I understand it.

> And the update will break the emulator for existing v25 users.

Sorry, which update do you mean, and why would it break for users with v25?

baaym commented:
Hi all,

On Windows slaves I'm unfortunately still running into issues. The plugin expects '.exe' files in the tools/bin directory, however they are '.bat' files in the currently latest Windows SDK: https://dl.google.com/android/repository/sdk-tools-windows-3859397.zip

> FATAL: Cannot run program "e:/jenkinsslave/tools/android-sdk/tools/bin/sdkmanager.exe": CreateProcess error=2, The system cannot find the file specified
> java.io.IOException: CreateProcess error=2, The system cannot find the file specified
> 	at java.lang.ProcessImpl.create(Native Method)
> 	at java.lang.ProcessImpl.<init>(ProcessImpl.java:386)
> 	at java.lang.ProcessImpl.start(ProcessImpl.java:137)
> 	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
> 	at hudson.Proc$LocalProc.<init>(Proc.java:245)
> 	at hudson.Proc$LocalProc.<init>(Proc.java:214)
> 	at hudson.Launcher$LocalLauncher.launch(Launcher.java:850)
> 	at hudson.Launcher$ProcStarter.start(Launcher.java:384)
> 	at hudson.Launcher$RemoteLaunchCallable.call(Launcher.java:1195)
> 	at hudson.Launcher$RemoteLaunchCallable.call(Launcher.java:1160)
> 	at hudson.remoting.UserRequest.perform(UserRequest.java:153)
> 	at hudson.remoting.UserRequest.perform(UserRequest.java:50)
> 	at hudson.remoting.Request$2.run(Request.java:336)
> 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
> 	at hudson.remoting.Engine$1$1.run(Engine.java:94)
> 	at java.lang.Thread.run(Thread.java:745)
> Caused: java.io.IOException: Cannot run program "e:/jenkinsslave/tools/android-sdk/tools/bin/sdkmanager.exe": CreateProcess error=2, The system cannot find the file specified
> 	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
> 	at hudson.Proc$LocalProc.<init>(Proc.java:245)
> 	at hudson.Proc$LocalProc.<init>(Proc.java:214)
> 	at hudson.Launcher$LocalLauncher.launch(Launcher.java:850)
> 	at hudson.Launcher$ProcStarter.start(Launcher.java:384)
> 	at hudson.Launcher$RemoteLaunchCallable.call(Launcher.java:1195)
> 	at hudson.Launcher$RemoteLaunchCallable.call(Launcher.java:1160)
> 	at hudson.remoting.UserRequest.perform(UserRequest.java:153)
> 	at hudson.remoting.UserRequest.perform(UserRequest.java:50)
> 	at hudson.remoting.Request$2.run(Request.java:336)
> 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
> 	at hudson.remoting.Engine$1$1.run(Engine.java:94)
> 	at java.lang.Thread.run(Thread.java:745)
> 	at ......remote call to JNLP4-connect connection from 10.69.4.230/10.69.4.230:50494(Native Method)
> 	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1545)
> 	at hudson.remoting.UserResponse.retrieve(UserRequest.java:253)
> 	at hudson.remoting.Channel.call(Channel.java:830)
> 	at hudson.Launcher$RemoteLauncher.launch(Launcher.java:959)
> 	at hudson.Launcher$ProcStarter.start(Launcher.java:384)
> 	at hudson.plugins.android_emulator.SdkInstaller.installComponent(SdkInstaller.java:218)
> 	at hudson.plugins.android_emulator.SdkInstaller.doInstall(SdkInstaller.java:97)
> 	at hudson.plugins.android_emulator.SdkInstaller.install(SdkInstaller.java:69)
> 	at hudson.plugins.android_emulator.AndroidEmulator.setUp(AndroidEmulator.java:259)
> 	at hudson.model.Build$BuildExecution.doRun(Build.java:157)
> 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:490)
> 	at hudson.model.Run.execute(Run.java:1735)
> 	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
> 	at hudson.model.ResourceController.execute(ResourceController.java:97)
> 	at hudson.model.Executor.run(Executor.java:405)

The 'bin' directory contents:

>  Directory of E:/jenkinsslave/tools/android-sdk/tools/bin
> 
> 09/07/2017  09:03    <DIR>          .
> 09/07/2017  09:03    <DIR>          ..
> 03/29/2017  00:46             2,223 archquery.bat
> 03/29/2017  00:46             2,837 avdmanager.bat
> 03/29/2017  00:46             2,211 jobb.bat
> 03/29/2017  00:46             3,336 lint.bat
> 03/29/2017  00:46             2,049 monkeyrunner.bat
> 03/29/2017  00:46             2,833 sdkmanager.bat
> 03/29/2017  00:46             2,182 uiautomatorviewer.bat
>                7 File(s)         17,671 bytes
>                2 Dir(s)  163,748,618,240 bytes free

Regards,

redeamer commented:
Hi,

@orrc: Sorry if I'm misleading. I try to be more specific: By update I mean update of the plugin: By 2.15 of the plugin for v25 (and up) the '-engine classic' flag is used. By current PR-71 code **no** '-engine' flag is used for <= 24 and >= 26, but the '-engine classic' is still set for v25 SDKs. If we now remove the flag for all versions, then for v25 users the emulator won't start after the update, cause the reason why it was added in the first place is still true for that users.
This is not an arguing for removal or non-removal, just wanted to clear it up, what I meant in the first place ;)
But still, if a notion is added in the Changelog, that v25 users needs to add '-engine classic' now manually as additional arguments, it's fine to remove it completely.

@baaym Thanks for the report. I was setting up a windows machine to test the behavior. I can (of course) reproduce the suffix-issue, but not the NPE you reported (but there is another issue with the config.ini parsing, which would have hit you later). It seems more ore less the same trace as previously reported and got fixed with commit b030902. Even if I'm pretty sure you used the latest commit as your own base, I just want to double-check.

I'm currently fixing a config file parsing issue on Windows: the config.ini of the AVD is treated as Properties in the plugin code, but the data is not stored in a Properties-compliant format. On windows there are '\' used to store the new config value 'image.sysdir.1', where on Linux the delimiter is ';'. As by definition single '\' are removed on reading the properties. Which breaks the content as soon as the plugin updates the config values.

redeamer commented:
Hi, now the extension-issue and the config-file parsing on Windows is fixed. Additionally all Unit-tests would run on Windows as well. @baaym the [latest build](https://ci.jenkins.io/job/Plugins/job/android-emulator-plugin/job/PR-71/8/artifact/target/android-emulator.hpi) works on my new set-up Windows 10-Slave, could you please retry, if the NPE still exists with that build.
regards

baaym commented:
Hi @redeamer,

I ran the latest build, and even though it now can find the batch files the `NullPointerException` still occurs:

> [android] No Android SDK found; let's install it automatically...
> Downloading and installing Android SDK from http://dl.google.com/android/repository/sdk-tools-windows-3859397.zip
> [android] Base SDK installed successfully
> [android] Going to install required Android SDK components...
> [android] Installing the 'platform-tool' SDK component(s)...
> $ e:/jenkinsslave/tools/android-sdk/tools/bin/sdkmanager.bat --include_obsolete platform-tools
> 
> $ e:/jenkinsslave/tools/android-sdk/tools/bin/sdkmanager.bat --list --verbose
> [android] Installing the 'tool,extra-android-m2repository,extra-google-m2repository' SDK component(s)...
> $ e:/jenkinsslave/tools/android-sdk/tools/bin/sdkmanager.bat --include_obsolete tools extras;android;m2repository extras;google;m2repository
> 
> FATAL: null
> java.lang.NullPointerException
> 	at hudson.plugins.android_emulator.util.Utils.getToolCommand(Utils.java:430)
> 	at hudson.plugins.android_emulator.util.Utils.runAndroidTool(Utils.java:476)
> 	at hudson.plugins.android_emulator.util.Utils.runAndroidTool(Utils.java:469)
> 	at hudson.plugins.android_emulator.util.Utils.runAndroidTool(Utils.java:463)
> 	at hudson.plugins.android_emulator.SdkInstaller.isPlatformInstalled(SdkInstaller.java:322)
> 	at hudson.plugins.android_emulator.SdkInstaller.installPlatform(SdkInstaller.java:268)
> 	at hudson.plugins.android_emulator.SdkInstaller.installDependencies(SdkInstaller.java:246)
> 	at hudson.plugins.android_emulator.AndroidEmulator.setUp(AndroidEmulator.java:271)
> 	at hudson.model.Build$BuildExecution.doRun(Build.java:157)
> 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:490)
> 	at hudson.model.Run.execute(Run.java:1735)
> 	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
> 	at hudson.model.ResourceController.execute(ResourceController.java:97)
> 	at hudson.model.Executor.run(Executor.java:405)

I guess the difference here is that I'm running this on a windows **slave**. Perhaps running on the master works fine.

Regards,

redeamer commented:
Hi @baaym,
my initial Windows-tests where on a slave-node too and I now changed the working dir of the windows slave to 'D:\jenkinsslave' to match your setup more closely (had some concerns that a non-C root may be a problem). But I still can't reproduce the NPE. I added some more log output, which may help to determine the reason of the crash you are facing.
If it would be possible, please rerun with the [latest build](https://ci.jenkins.io/job/Plugins/job/android-emulator-plugin/job/PR-71/9/artifact/target/android-emulator.hpi) post the output again, there should be now one more log-line which should give a clue, why the getAndroidSdk returns null (which should be the problem, deduced from the last crash-output).
Of course I would appreciate if anyone else with Windows-machines may take the time to test it there, to maybe gather more information.
Regards

illumi commented:
On Windows when running Jenkins as a service, AVDs seem to be created in `C:\Windows\system32\config\systemprofile\.android\avd\`. The AVD seems to be successfully created in this location but both `emulator.exe -list-avds` and `android.bat list avd` show no AVD could be found which causes the plugin to fail.

orrc commented:
@illumi Might this be the issue with "Local system" accounts?
https://wiki.jenkins.io/display/JENKINS/Android+Emulator+Plugin#AndroidEmulatorPlugin-Knownissues

desaibhavik commented:
Hi Michael

We also tried the latest version which works for android-23 and android-24 with qemu2
But I am unable to get android-26 working when Target ABI is specified 

1. "google_apis;x86" we see error "Failed to find package system-images;android-26;default;google_apis;x86"
2. "x86" we see error "Failed to find package system-images;android-26;default;x86"

Default tag is always added by default. Is there any workaround for this?

redeamer commented:
Hi,
I've made some updates to the [latest build](https://ci.jenkins.io/job/Plugins/job/android-emulator-plugin/job/PR-71/12/artifact/target/android-emulator.hpi) of the plugin.
- Fix issue with sdkmanager, that x86 images where not installed if the x86_64 was already present
- Fix issue, that the plugin was expecting avdmanager and sdkmanager be present on all v25 SDK Tools, but this is only true for SDK Tools >= v25.3
- Fix issue if a tag is specified in the ABI-field for the new tools, this works now
- On system with v26.0.1 and lower the avdmanager expects --tag, even if only a single image is present, no the tag is always specified (except 'default')
- If a preview-sdk license needs to be accepted, the plugin will handle that now
- It is now optional (also with the avdmanager command) to specify the device definition, the plugin handles the 'user input' needed, if no device definition is given.

@desaibhavik: Thanks for the report, due to them I could fix some issues with a 'tag' given with the ABI. For the ABI field: if you specify a tag it has to be in the form `<tag>/<abi>` eg: `google_apis/x86`. I made this more clear in the description. Currently no 'default' tag is present for android-26 images, so it is mandatory to specifiy a tag (in this case google_apis). Anyways you would need to update to the latest build for that to work.
@baaym: any updates on your issue?

regards

redeamer commented:
Hi,
I've made some updates reflected by the [latest build](https://ci.jenkins.io/job/Plugins/job/android-emulator-plugin/job/PR-71/13/artifact/target/android-emulator.hpi) of the plugin.
- Fix issue with sdkmanager: x86 images where not downloaded, if the x86_64 was already present
- Fix issue, that the plugin was expecting avdmanager and sdkmanager be present on all v25 SDK Tools, but this is only true for SDK Tools >= v25.3
- Fix issue if a tag is specified in the ABI-field for the new tools
- On system with v26.0.1 and lower the avdmanager expects --tag, even if only a single image is present, now the tag is always specified (except 'default')
- If a preview-sdk license needs to be accepted, the plugin will handle that now
- It is now optional (as with the avdmanager command) to specify the device definition, the plugin handles the 'user input' needed, if no device definition is given.

@desaibhavik: Thanks for the report, due to this I could fix some issues with a 'tag' given with the ABI. For the ABI field: if you specify a tag it has to be in the form `<tag>/<abi>` eg: `google_apis/x86`. I made this more clear in the description. Currently no 'default' tag is present for android-26 images, so it is mandatory to specifiy a tag (in this case google_apis). Anyways you would need to update to the latest build for that to work.
@baaym: any updates on your issue?

regards Michael

redeamer commented:
Hi @orrc,
- I finally removed the `-engine classic`, even for v25. So this needs to be added to the Known Issues and addressed in the changelog.
- I mentioned the 'updating of the sdk' in the past: You convinced me, that we should be careful and not simply call 'sdkmanager --update'. But I would suggest another approach to that topic: If the 'Automatically install Android components when required' is set the plugin should download the sdk tools if the version is older than the version the plugin would initially install (by now 26.0.1). So we're still sure, that the plugin is compatible with the commands, and we only need to handle one version of all of the sdkmanager commands.
- off-topic: Do you plan on some point, to fade out deprecated topics? Like unsupported android target versions, the legacy ant project support... ?
br Michael

orrc commented:
Hey, sorry I didn't reply to your previous reply…

You were right that we should keep `-engine classic` for the specific 25.x versions of the SDK that didn't support `-ports` and whatever else that was required.  I had somehow forgotten about these broken SDK tools versions… :/

The idea about "Auto-install when required" is probably also reasonable, though there's still a risk that we update to a "too new" version.  But then we should try and fix this plugin if breakage occurs.  So yeah, sounds good.

And yeah, the Ant stuff can definitely be marked as deprecated; I should do that at some point in the future.

Nice work on all the fixes! And thanks everyone for testing!

baaym commented:
Hi @redeamer,

It seems the plugin now works fine on a Windows slave - also when not running as a local system user. Thanks for these fixes!

Regards,

redeamer commented:
Hi @orrc,
so, engine classic for 25.x is still set. I added the feature, that if we auto-install the SDK and the tools is outdated (= version < than version we would install if nothing found at all (26.0.1 for now)), the tools are updated to the plugins default version. (I hope the behavior is clear from the description).

Additionally I refactored the getAndroidSdk method and merged the discoverAndroidHome into the method, as they were both validating SDK directories. discoverAndroidHome was validating all potential directories except checking the path. The valid dir was returned and passed to getAndroidSDK, where it was validated again, and if discoverAndroidHome returned null, getAndroidSDK was validating the directories from the path.
So for now, getAndroidSDK is building up a list of all possible directories (including PATH extracted dirs), and then checks them in the given order (same order as before the change). If a valid dir is found, the AndroidSdk is returned directly. So there is one validation call less, and imho readability got increased.

@baaym: Awesome, thanks for your feedback.

regards michael

orrc commented on src/main/java/hudson/plugins/android_emulator/sdk/cli/SdkToolsCommandsCurrentBase.java:
```diff
@@ -13,22 +14,27 @@
+            final String modifiedComponent = component
+                    .replaceAll("^tool$", "tools")
+                    .replaceAll("^platform-tool$", "platform-tools")
+                    .replaceAll("^addon-$", "addons-")
```
The `$` here looks suspicious.

redeamer replied:
Nice spotting, thanks. Good that there is line 27 doing it properly.
I'll remove the line on next commit.

redeamer replied:
Removed with last commit. The existing `replaceAll("^addon-", "add-ons;addon-")` covered it already correctly.

orrc commented:
> I refactored the getAndroidSdk method and merged the discoverAndroidHome into the method

Yeah, sounds about right — I'm sure I kept wondering why I (probably?) made them separate, every time I looked at the codebase 😄 

I'll try and give this a good looking at in the next day or so!

redeamer commented:
Removed the unneeded line. No effective change to plugin behavior.

bperel commented:
Is there any change that this PR can make it for a merge?

redeamer commented:
Hi @orrc, any more findings/comment/suggestions on the PR? I would have some time next week to work on it. regards

orrc commented:
Hey @redeamer — yeah, I was testing some more earlier this week, and things generally look good.

It doesn't seem possible to create emulators for API <= 14 with `avdmanager`, but that's an Android bug, and shouldn't block release, so long as we mention it.  I don't know if we can fall back to `android create avd` in those cases, but that's a matter for another PR anyway.

I believe I also found some usage of a Java 8 API (something like `process.isAlive()`), but I _think_ that was in existing code.

I'll give the open PRs more attention at the weekend and then we should be able to release within the next week… :tada:

redeamer commented:
Hi @orrc, good to hear!
I just tried an android-10 image, it was working... maybe there is something else going on. I will look more closely into it tomorrow.
The `process.isAlive()` call was newly added, I'll replace the the call by a < Java 8 API version.

orrc commented:
> I just tried an android-10 image, it was working

Yeah, weirdly, android-10 works, but the others don't.

Having said that, inside a Docker container, if I make this diff, I can get the emulators to at least be created (doesn't work on my MacBook itself, for some reason):
```diff
diff --git a/src/main/java/hudson/plugins/android_emulator/Constants.java b/src/main/java/hudson/plugins/android_emulator/Constants.java
index 4aa8aa1..81341e0 100644
--- a/src/main/java/hudson/plugins/android_emulator/Constants.java
+++ b/src/main/java/hudson/plugins/android_emulator/Constants.java
@@ -356,6 +356,10 @@ class AndroidPlatform implements Serializable {
     }

     public String getPackagePathOfSystemImage(final String abi) {
+        if (!requiresAbi()) {
+            return String.format("platforms;android-%d", getSdkLevel());
+        }
+
```

> The `process.isAlive()` call was newly added

Oh, cool, thanks!

redeamer commented:
the `requiresAbi` was the correct hint. If the requiresAbi returns false (eg. android-14), no system image is going to be installed. At first sight, this seems to be the same behavior as in the original code.

For a proper fix I was looking more closely to the different system image names for the different tools versions (sysimg-xxx, sys-img-xxx, system-image;xxx) and identified the following (I would love, if this could be re-verified by another person):
- It is not possible anymore for sdk-tools r21 and below to install new images and platform-tools
- For sdk-tools r22 it is possible to install platform-tools, but no images.

If that's true, the requiresAbi check on installation can definitely be dropped, as sdk-tools r22 and up, using consistent image names, all with abi and tag.

redeamer commented:
Hi @orrc,
one question about `SdkInstaller.installDependencies` (master and this branch):

There are the lines
`boolean requiresAbi = !emuConfig.isNamedEmulator() && emuConfig.getOsVersion().requiresAbi();
String abi = requiresAbi ? emuConfig.getTargetAbi() : null;`

As I see it right, the intention to have the `namedEmulator` influencing `abi` here, is to prevent the system-image installation?

As I cannot test the behavior of the old sdk-tools, because it seems that they cannot download the images anymore (see previous post), maybe you could shed some light why requiresAbi was added to the download of images, and if it is still needed (which I think not).

regards

orrc commented:
Originally, there were no system images — the platform images themselves were always armeabi.  For API 10, Intel created a separately downloadable x86 image, and the ABI field got added to the config of this plugin, so that it could be passed when calling `android create avd`.

This plugin would ignore the ABI field if it realised that the configured API version didn't support ABIs — that's the `requiresAbi()` method which still exists.  This allows you to create a matrix job with various API versions, specifying "x86" as the ABI, and the AVDs will still be successfully created, even if they don't need an ABI.  That was especially useful before Google started offering/hosting x86 images, which started with API 14 or 15.

But yeah, the `requiresAbi()` check probably wasn't really required in the past during installation, as you mention — and as [the current comment][0] mentions.

However, with the current `sdkmanager`, ignoring the `requiresAbi()` check and attempting to download a system image for older platform versions will fail.
e.g. If I alter `requiresAbi()` to always return `true` and attempt to create an emulator for API 2–9 or 11–13:

```
Building remotely on localhost-a14290f93021 (android-26.0.1) in workspace /home/jenkins/workspace/android_4.0
$ /opt/android/sdk/tools/bin/avdmanager list target
[android] The configured Android platform needs to be installed: android-9
$ /opt/android/sdk/tools/bin/avdmanager list target
$ /opt/android/sdk/tools/bin/avdmanager list target
[android] Installing the 'android-9,sys-img-armeabi-v7a-android-9' SDK component(s)...
$ /opt/android/sdk/tools/bin/sdkmanager --include_obsolete platforms;android-9 system-images;android-9;default;armeabi-v7a
Warning: File /home/jenkins/.android/repositories.cfg could not be loaded.
Warning: Failed to find package system-images;android-9;default;armeabi-v7a
```

But if I keep the `requiresAbi()` check, and apply [the patch][1] in my earlier comment, I can create an AVD for these platform versions that didn't have separate system images.

I'm not sure whether that helps to answer your question? 😄

About your previous comment, I wasn't sure why older SDK Tools versions (21, 22 etc.) were relevant.  Either users have the "auto-install" flag enabled, so this plugin will now auto-update to r26.0.1 and we use the now-consistent `system-images;android-X;…` format (or we don't use it for older API levels) — or they have that flag disabled, and it's up to them to manually install the correct components, and the plugin doesn't need to care.

In any case, I'm trying to find out whether Android even intend to support older API versions with the SDK Tools, as I note that creating an `android-9` AVD from `avdmanager` works in r26.0.1, but it throws an exception with newer versions… 🤔

[0]:https://github.com/jenkinsci/android-emulator-plugin/blob/android-emulator-2.15/src/main/java/hudson/plugins/android_emulator/SdkInstaller.java#L343
[1]:#issuecomment-341928062

redeamer commented:
Hi @orrc,
thanks for clarification about the historical sys-img behavior.

```About your previous comment, I wasn't sure why older SDK Tools versions (21, 22 etc.) were relevant.```
Sorry for the confusion, you are right, they aren't relevant. I had a different behavior in mind (auto-update only if autoInstallDir (<jenkinsroot>/tools/android-sdk), which would maybe the clearer solution?). I will add a thought in the end of the comment the auto-install.

About the android-9, there are no system-images listed in for those versions: there are only images for: android-10, android-14 (only arm-v7a) and above. So this should be fine, that the user gets an error here. We could of course do some basic check in the code, but I don't thinks that is necessary.

Additional thought about auto-install:
I personally think it would be much clearer
- A: if auto-install is selected, the directory entry in the config should be grayed and the plugin should always use '<jenkinsroot>/tools/android-sdk' and only if auto-install is not used: the other paths are used (user-defined path, common ENV variables, PATH-check).
- or B: if auto-install is selected, and a user-configured path is set, we will always use this, and only if empty the '<jenkinsroot>/tools/android-sdk'
And only use the enviroment and path to detect sdk's which are never handled by Jenkins.
Otherwise we could manipulate a directory, never intended to be changed, just because it was on the path.

FnTm commented:
Hey @orrc @redeamer just curious - how much work is there left for this PR to be merged into the master? This functionality is something that I'd really like to use in my Jenkins ASAP.

redeamer commented:
I untangled the requiresAbi (even if not really needed) and the isNamedEmulator in SdkInstaller.
Now if a named emulator is used, the loading of the system-image is explicitly suppressed (and not via the abi != null). Additionally the requiresAbi check is ignored for new SDKs, so the ABI is used all the time with newer SDKs (as all images available images have them now inserted).
With those changes, both android-10 and android-14 can be loaded an created.

redeamer commented:
Hi @orrc, me again, ahh, now I got it (I somehow missed your bugreport, but as I read it, it's clear), the image was shipped with the platform with older versions, and not separately.
Okay, now I can fix it properly, should work by tomorrow.

redeamer commented:
Hi, I've updated the plugin code, that for versions !requiresAbi the platforms package is used. Now android-9 targets can be created (tested android-9, android-10, android-11, android-14 and android-24), at least with r26.0.1.

Additional Info: the requiresAbi was checking for `android-10 || >= android-15`, but current android-14 does not ship an image in platform, just explicitly (system-images;android-14;default;armeabi-v7a). I changed the check to `android-10 || >= android-14`. I'm not sure if there was a change by google and we now break creation if an 'old' android-14 platform was loaded (if it was integrate), or this was wrong in the plugin.

redeamer commented:
Hi @orrc, any more input/suggestions/...?

orrc commented:
No, I'll merge it now and see what else I can get in before a release! 🎉

Thanks a lot for all your work and testing and fixing on this!
